src/msg/async/AsyncConnection.cc

```diff
@@ -967,17 +967,31 @@ int AsyncConnection::_process_connection()
           ::close(sd);
         }
 
-        sd = net.connect(get_peer_addr());
+        sd = net.nonblock_connect(get_peer_addr());
         if (sd < 0) {
           goto fail;
         }
-        r = net.set_nonblock(sd);
+        center->create_file_event(sd, EVENT_READABLE, read_handler);
+        first_connect = true;
+        state = STATE_CONNECTING_RE;
+        break;
+      }
+
+    case STATE_CONNECTING_RE:
+      {
+        r = net.reconnect(get_peer_addr(), sd);
         if (r < 0) {
+          if (first_connect == true) {
```

Comment:
“2015-09-06 13:31:51.563489 7fcde1ffb700 20 -- :/1551015 >> 172.16.50.12:6789/0 conn(0x7fcde4171090 sd=18 :-1 s=STATE_CONNECTING_RE pgs=0 cs=0 l=1).process state is STATE_CONNECTING_RE, prev state is STATE_CONNECTING
2015-09-06 13:31:51.563497 7fcde1ffb700 10 NetHandler reconnect reconnect: Operation already in progress”
Subsquent calls to reconnect may return EALREADY, so we also check it. And let it to break?